### PR TITLE
Chrome-screenshot-based rendering for WebPageReportSources

### PR DESCRIPTION
**Pull request by speezepearson**
- From [a much earlier attempt](https://github.com/speezepearson/metrics-portal/compare/master...speezepearson:report-screenshot-email?expand=1), adapt code that wraps a Chrome devtools library.
- Manually tested to ensure that Guicing/rendering worked.
- Removed `BaseScreenshotRenderer` class, since there wasn't enough shared code between HTML and PDF for it to be worthwhile.

### CODE REVIEW

File: conf/portal.application.conf [hunk 1]
```diff
@@ -147,9 +147,11 @@ reporting {
     WEB_PAGE {
       "text/html" {
         type = "com.arpnetworking.metrics.portal.reports.impl.chrome.HtmlScreenshotRenderer"
+        chromePath = "" # will need to be changed to the actual Chrome executable path in order to use this renderer
```
Comment (vjkoskela) [inline conf/portal.application.conf]: will this respect `$PATH`?

Comment (speezepearson) [reply conf/portal.application.conf]: Nope! This option must be given and it must point to the Chrom[e/ium] executable, or else an exception will get thrown whenever a web-based report tries to render.

File: README.md [hunk 2]
```diff
@@ -109,6 +109,10 @@ For example:
 
 Arguments before the `--` are interpreted by the JVM while arguments after `--` are passed to Metrics Portal.
 
+##### Extras
```
Comment (vjkoskela) [inline README.md]: I think rename this section to Reporting and please mention if using the published Docker image that this is available by default and how one configures the chromium path to use it.

File: app/com/arpnetworking/metrics/portal/reports/impl/chrome/DevToolsFactory.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                new ChromeLauncher.RuntimeShutdownHookRegistry(),
+                new ChromeLauncherConfiguration()
+        );
+        return launcher.launch(true);
+
+    }
+
+    private final String _chromePath;
+
+    private static final Map<String, ChromeService> CHROME_SERVICE_BY_PATH = Maps.newHashMap();
```
[the remark this answers is not in the document]
Comment (speezepearson) [reply app/com/arpnetworking/metrics/portal/reports/impl/chrome/DevToolsFactory.java]: Yep, thanks!

File: app/com/arpnetworking/metrics/portal/reports/impl/chrome/DevToolsService.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * A relatively minimal interface for a Chrome tab's dev tools.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public interface DevToolsService {
+    /**
+     * Evaluates some JavaScript in the tab.
+     *
+     * @param js A JavaScript expression. (If you need multiple statements, wrap them in an IIFE.)
```
[the remark this answers is not in the document]
Comment (speezepearson) [reply app/com/arpnetworking/metrics/portal/reports/impl/chrome/DevToolsService.java]: https://developer.mozilla.org/en-US/docs/Glossary/IIFE

File: main/docker/Dockerfile [hunk 5]
```diff
@@ -69,15 +72,18 @@ ENV APP_OPTS="/opt/metrics-portal"
...
     mkdir -p /opt/metrics-portal/config && \
     mkdir -p /opt/metrics-portal/logs
 ADD deps /opt/metrics-portal/lib/
 ADD bin /opt/metrics-portal/bin/
 ADD config /opt/metrics-portal/config/
 ADD lib /opt/metrics-portal/lib/
+RUN chown -R mportal /opt/metrics-portal
 
 # Entry point
+USER mportal
```
Comment (vjkoskela) [inline main/docker/Dockerfile]: What does this do to log file permissions generated by the application when they are volume mounted out of the running container? 

Note: the behavior may be different between Docker for Mac and Docker for Linux; the former uses a VM under the covers. Testing the latter could be done by starting a VM and running the Docker image in there.

File: test/java/com/arpnetworking/metrics/portal/reports/impl/HtmlScreenshotRendererIT.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * Tests class {@link HtmlScreenshotRenderer}.
+ *
+ * This test-class is only meant to be run manually: it depends on Chrome, which not all environments are guaranteed to have, so this class
+ * is marked {@code @Ignore}. If you want to run it manually, set {@link #CHROME_BINARY_PATH} as appropriate for your system first.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+@Ignore
```
Comment (vjkoskela) [inline test/java/com/arpnetworking/metrics/portal/reports/impl/HtmlScreenshotRendererIT.java]: Does this still need to be ignored?

Comment (speezepearson) [reply test/java/com/arpnetworking/metrics/portal/reports/impl/HtmlScreenshotRendererIT.java]: Resolved in person. Tldr, `@Ignore`ing this is fine for now, but we should add a real integration test at some point. Tracked in Jira.

File: conf/portal.application.conf [hunk 7]
```diff
@@ -144,11 +144,25 @@ kairosdb {
...
   renderers {
     WEB_PAGE {
-      "text/html" {
+      "text/html; charset=utf-8" {
         type = "com.arpnetworking.metrics.portal.reports.impl.chrome.HtmlScreenshotRenderer"
+        chromePath = "/usr/lib/chromium/chrome"
+        # ^ This path is only correct in the MPortal Docker image! If you aren't running MPortal in Docker, and you have reports enabled,
+        #   you'll need to change this to the actual path to your Chrome binary.
+        chromeArgs = {
+          # Command-line flags passed to Chrome. List of valid args: https://peter.sh/experiments/chromium-command-line-switches/
```
Comment (vjkoskela) [inline conf/portal.application.conf]: Are all chrome arguments key-value? Some programs take "flags" as well which have no associated value. How do you represent these here?

File: app/com/arpnetworking/metrics/portal/reports/impl/chrome/DevToolsServiceWrapper.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+/**
+ * Wraps a {@link com.github.kklisura.cdt.services.ChromeDevToolsService} to conform to the {@link DevToolsService} interface.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public class DevToolsServiceWrapper implements DevToolsService {
+    private final com.github.kklisura.cdt.services.ChromeDevToolsService _dts;
+
+    /**
+     * @param dts The {@link com.github.kklisura.cdt.services.ChromeDevToolsService} instance to wrap.
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/reports/impl/chrome/DevToolsServiceWrapper.java]: nit: incomplete javadoc

**Pull request merged**